apps/src/lib/kits/maker/boards/microBit/LightSensor.js

```diff
@@ -60,18 +77,76 @@ export default class LightSensor extends EventEmitter {
 
   start() {
     this.board.mb.enableLightSensor(); // enable light sensor
-    this.board.mb.streamAnalogChannel(sensor_channels.lightSensor);
+    this.board.mb.streamAnalogChannel(SENSOR_CHANNELS.lightSensor);
   }
 
   stop() {
-    this.board.mb.stopStreamingAnalogChannel(sensor_channels.lightSensor); // disable light sensor
+    this.board.mb.stopStreamingAnalogChannel(SENSOR_CHANNELS.lightSensor); // disable light sensor
+  }
+
+  // Reset the state to initial values between runs
+  reset() {
+    this.state = {
+      threshold: 128,
+      rangeMin: 0,
+      rangeMax: 255,
+      currentReading: 0,
+      currentBufferWriteIndex: 0
+    };
+    this.state.buffer.fill(0);
   }
 
-  //TODO
-  getAveragedValue() {}
+  // Get the averaged value over the given ms, adjusted within the range, if specified.
+  // If ms is outside of 50 and 3000 range, prints a warning to debug console.
+  getAveragedValue(ms) {
+    let opts = {ms};
+    apiValidateTypeAndRange(
+      opts,
+      'lightSensor.getAveragedValue',
+      'ms',
+      opts.ms,
+      'number',
+      50,
+      3000
+    );
+
+    // Divide ms range by sample rate of sensor
+    let requestedRange = Math.ceil(ms / SAMPLE_INTERVAL);
+    let indicesRange = requestedRange;
+    // User requested average over greater range than is recorded, so average over all
+    // recorded data.
+    if (requestedRange >= this.state.currentBufferWriteIndex) {
+      indicesRange = this.state.currentBufferWriteIndex;
+    }
+    let endIndex =
+      this.state.currentBufferWriteIndex % this.state.buffer.length;
```

Comment:
Should for loops be inclusive? I noticed you also suggested that earlier in this code review. Is it a best practice?

Reply:
It comes down to the situation.  It's common when doing a 0-based count to use `<` in the condition , but in a situation like this, where we are very precisely iterating over an inclusive range, I think it's nice to use `<=` so that we can see exactly which indices we are processing.  If the condition were to compare against a `count` then `<` would read correctly, but given that we're talking about `endIndex` I think `<=` makes a bit more sense.

Reply:
Huh, cool. TIL. Updating . . .